### PR TITLE
fix(clustering): DP friendly msg when CP offline

### PR DESCRIPTION
**Pull request by StarlightIbuki**
Translate the log to a more user-friendly form, and lower the log level.

Also correcting some of the log formats.

Fix FT-3204

### CODE REVIEW

File: kong/clustering/data_plane.lua [hunk 1]
```diff
@@ -262,11 +264,18 @@ function _M:communicate(premature)
   ngx.thread.kill(write_thread)
   c:close()
 
+  local true_err
+  
   if not ok then
+    true_err = err
+  else
```
Comment (chronolaw) [inline kong/clustering/data_plane.lua:271]: style: blank line before `else`

File: kong/clustering/data_plane.lua [hunk 2]
```diff
@@ -262,11 +264,18 @@ function _M:communicate(premature)
   ngx.thread.kill(write_thread)
   c:close()
 
+  local true_err
```
Comment (chronolaw) [inline kong/clustering/data_plane.lua:267]: how about `true_err` => `err_msg`

File: kong/tools/utils.lua [hunk 3]
```diff
@@ -1468,4 +1468,8 @@ do
 end
 _M.time_ns = time_ns
 
+function _M.endswith(s, e)
```
Comment (chronolaw) [inline kong/tools/utils.lua:1471]: shall we use `pl.stringx.endswith` ?

File: kong/clustering/data_plane.lua [hunk 4]
```diff
@@ -262,11 +265,19 @@ function _M:communicate(premature)
...
 
-  elseif perr then
-    ngx_log(ngx_ERR, _log_prefix, perr, log_suffix)
+  else
+    err_msg = perr
+  end
+  
+  if endswith(err_msg, ": closed") then
+    ngx_log(ngx_INFO, _log_prefix, "connection to control plane closed", log_suffix)
+  elseif err_msg then
```
Comment (chronolaw) [inline kong/clustering/data_plane.lua:279]: style: blank line before `elseif`

File: kong/clustering/wrpc_data_plane.lua [hunk 5]
```diff
@@ -220,11 +222,20 @@ local function communicate_impl(dp)
...
-  elseif perr then
-    error(perr)
+  else
+    err_msg = perr
+  end
+  
+  if endswith(err_msg, ": closed") then
+    ngx_log(ngx_INFO, _log_prefix, "connection to control plane closed", log_suffix)
+    return
+  elseif err_msg then
```
Comment (chronolaw) [inline kong/clustering/wrpc_data_plane.lua:237]: style: blank line before `elseif`

File: spec/02-integration/09-hybrid_mode/02-start_stop_spec.lua [hunk 6]
```diff
@@ -132,3 +132,38 @@ for cluster_protocol, conf in pairs(confs) do
     end
   end)
 end
+
+-- note that lagacy modes still error when CP exits
+describe("when CP exits before DP", function()
+  local need_exit = true
+  setup(function()
```
Comment (chronolaw) [inline spec/02-integration/09-hybrid_mode/02-start_stop_spec.lua:139]: style: we should add a blank line to separate each function.

**Review by chronolaw — APPROVED** [review]

File: kong/clustering/data_plane.lua [hunk 7]
```diff
@@ -262,11 +265,20 @@ function _M:communicate(premature)
   ngx.thread.kill(write_thread)
   c:close()
 
+  local err_msg
```
Comment (dndx) [inline kong/clustering/data_plane.lua:268]: You can just write `local err_msg = ok and err or perr` instead. Might be easier to read.

**Review by fffonion — APPROVED** [review]

**Review by dndx — APPROVED** [review]

**Pull request merged**